Add new udc driver support for ELAN IC

sywang-emc opened the pull request:
Add new udc driver support for ELAN IC.

Depends on https://github.com/zephyrproject-rtos/zephyr/pull/97843

sywang-emc committed:
Add new udc driver support for ELAN IC

dsseng commented on drivers/usb/udc/udc_elan_elandev.c:
```diff
@@ -0,0 +1,1698 @@
+		ep = priv->ep_out_num_new;
+	}else {
+		udc_buf_put( cfg, buf);
+	}
```
Please format the code here according to the code style of Zephyr code base

dsseng reviewed (commented):
Thank you for your contibution! Could you please fix the warnings created by the CI, as well as add your signed-off-by to the commit to certify Developer Certificate of Origin?

These changes would help accelerate review of your PR. Thanks!

sywang-emc committed:
format code as zephyr style

Signed-off-by: sywang emc <atlantis.wang@emc.com.tw>

jfischer-no reviewed (changes requested):
@sywang-emc You need to get familiar with https://docs.zephyrproject.org/latest/contribute/style/code.html and https://kernel.org/doc/html/latest/process/coding-style.html. Also, please get familiar with the code and style in USB area. Please fix all the style issues in your code. There should not be any typedefs. You need to learn to write meaningful commit messages. For that, you can look at the git log or merged pull request in the USB area. Please describe who is the controller designer and what controller type it is.

sywang-emc committed:
drivers: usb: udc: support for USB on elan em32f967 series soc

This driver support usb controller on elan em32f967 series soc, and it is written base on udc helper functions by Shunyi Wang. It adopts the definition from em32f967 HAL, but does not use any function from it. Some patches are applied to meet requirements of a udc driver, without those patches, it cannot work properly as a udc driver.

Signed-off-by: sywang emc <atlantis.wang@emc.com.tw>

sywang-emc committed:
drivers: usb: udc: support for USB on elan em32f967 series soc

This driver support usb controller on elan em32f967 series soc, and it is written base on udc helper functions by Shunyi Wang. It adopts the definition from em32f967 HAL, but does not use any function from it. Some patches are applied to meet requirements of a udc driver, without those patches, it cannot work properly as a udc driver. Fix software qualities impacted issues.

Signed-off-by: sywang emc <atlantis.wang@emc.com.tw>

dsseng reviewed (commented):
Some suggestions, feel free to disagree with those if they do not fit

dsseng reviewed (commented):
Also please add a DT binding for the new compatible and fix minor style issues like extra newlines. `clang-format` is one of the recommended tools and it would be great to use it on newly added code

dsseng commented:
> I will follow your suggestions and examine the code more thoroughly. By "DT binding," do you mean adding the ELAN UDC driver's YAML file to the zephyr\dts\bindings\usb directory, or the board DTS? If it's the board DTS, it's currently under review.

I mean yaml, the binding file that includes information on the required properties.

I'm not sure if it's a must for merging the driver, but am inclined to think so since it's necessary to ensure the proper usage of the driver.

Please add the binding in a separate commit, see the message format in the other DT bindings.
